### ISSUE TITLE
Borderless Window features

### ISSUE THREAD

**Issue by sirgru** [glfw/glfw#2266]
It would be good if GLFW implemented features on an undecorated window:
1. Draggability on a custom screen area. Support for standard features like window snapping on Windows.
2. Resizing.
3. APIs for minimize/maximize/fullscreen.

These issues are summarized here: https://discourse.glfw.org/t/immovable-resizable-bordeless-window/1586

These features are required because the titlebar is not customizable. Most modern applications will want to reuse the titlebar. Windows supports complete customization of the titlebar in WPF, yet the one drawn with GLFW is old style which doesn't even allow for color customization.

IIRC Electron has this functionality and Flutter does too: https://github.com/bitsdojo/bitsdojo_window

This makes GLFW lag behind modern frameworks. Most of these features seem to be implemented a long time ago for Windows and Linux, and were never merged. It would be good if these features could be merged, because this would significantly modernize the framework for non-games applications.

**Comment by mdrejhon** [comment]
Ditto.  

However, keep support for fullscreen exclusive (e.g. where GLFW siezes exclusive control of the screen on some platforms).

That's required for some applications (e.g. the VSYNC OFF / beamracing ideas posted in #1157).  

It also reduces lag/stutter on different-Hz multimonitor systems, since Microsoft Windows is a single-Hz compositor (which means smooth graphics on on screen, stuttery graphics on the other screen) when running in windowed mode or borderless windowed mode on a multiple-Hz system (e.g. 60Hz monitor + 144Hz monitor).

So in a nutshell, all of these need to be support in as cross-platform manner:

* Windowed
* Boderless full screen window *(occasionally-compromised hybrid of windowed and fullscreen)*
* Full screen exclusive *(lowest-lag, highest-performance, least-stutter, most control)*

Most platforms supports at least 2 out of 3, and several supports 3 out of 3.

**Comment by dougbinks** [comment]
There is already a preexisting issue open for support for resizing undecorated windows #990

**Comment by Seneral** [comment]
True, but it's dissappointing to see there has been no action taken on any of the many merge requests. What do you think does it take for glfw to finally merge these? As a developer, I really hate having to scratch together solutions from many merge requests to get working what should long have been merged.
Currently, I only have dragging working, snapping on windows doesn't work (but on X11/Wayland), and resizing doesn't work, still haven't gotten around to grab that from any one of the many PRs.

Seriously, we need this for proper desktop application support, please do something about it, there are plenty people willing to work on this but seemingly no commitment by anybody able to merge.

**Comment by dougbinks** [comment]
@Seneral 

>True, but it's dissappointing to see there has been no action taken on any of the many merge requests. What do you think does it take for glfw to finally merge these? As a developer, I really hate having to scratch together solutions from many merge requests to get working what should long have been merged.

>Seriously, we need this for proper desktop application support, please do something about it, there are plenty people willing to work on this but seemingly no commitment by anybody able to merge.

GLFW isn't a commercial product which owes it's customers a certain level of service. It's primarily maintained by one person for no pay. Indeed on top of the time needed development of GLFW costs money due to the wide range of hardware required.  Comments such as yours do not help to motivate the maintainers. PRs are not merged by simply waiving a wand, and there's a huge amount of unseen work which goes into the development of GLFW.

**Comment by Seneral** [comment]
I didn't know it was one person - maybe it's time to take on more maintainers then. Is there anything I can do to make it easier to vet these merge requests, which people already seem to be using in production? Some of them have been open since 2017, so I really think it's time to do something. If the current process allows merge requests to stay open for 6 1/2 years (which wastes everybodys time), that's not cool.
Again, my intent was not to imply we deserve support - in good open source spirit there have been people giving solutions from quite early on, I just see no indication this is taken seriously at all, which leaves me concerned, especially since this is required to support e.g. wayland on gnome, where there are NO ways to drag or resize a glfw window right now (had to disable libdecor integration of glfw because it hard-crashes my application, probably separate issue).
I just hate to see that the only real solution seems to be "make a fork and merge it yourself" - which means tons of duplicate work and wasted time.

**Comment by dougbinks** [comment]
> I didn't know it was one person - maybe it's time to take on more maintainers then.

Taking on maintainers is not easy and doesn't always solve things.

>  If the current process allows merge requests to stay open for 6 1/2 years (which wastes everybodys time), that's not cool.

There are always multiple different priorities for different people. Simply keeping old features from working is a lot of work: supporting an evolving Wayland, dealing with new OS features which break things, or MacOS' deprecation of older features we rely on. As I mentioned even merging working PRs which cover the entire array of platforms isn't as simple as clicking 'merge'. There is usually a lot of work to do to with any PR.

**Comment by Seneral** [comment]
For the record, #2152 is, IMO, ready to merge

**Comment by dougbinks** [comment]
It takes a considerable amount of time for us to review and merge a PR, even one which are *ready to merge*. Even simple ones with only one OS affected and a very small amount of code take a considerable amount of time to check and merge..

**Comment by Seneral** [comment]
I didn't write it with the intend to argue - that's what the "for the record" was for. I don't expect anything, don't worry

**Comment by p0ryae** [comment]
Any updates? It's been a while.

I decided to do a full review of PR #2152 and ran into a Wayland-specific bug: a segfault when dragging a window. I debugged it, identified the root cause, and implemented a fix. I also tested the PR on X11 and Windows, and in my tests everything works as expected.

Details of the bug and my fix: https://github.com/glfw/glfw/pull/2152#issuecomment-3587387211

If you'd like, I can open a new PR with my set of change. I'm willing to follow through as many steps as necessary.

**Comment by dougbinks** [comment]
> Any updates? It's been a while.

No, we have a large backlog of issues / PRs and it will be some time before issues like this can be addressed.

**Comment by p0ryae** [comment]
> No, we have a large backlog of issues / PRs and it will be some time before issues like this can be addressed.

No worries. @envyen kindly added me as a collaborator to his repo to include my improvements, and I’ve pushed the change to the repo in order to reflect it on the PR #2152. We are both committed to follow up as necessary.
